lib/Cryptol.cry

```diff
@@ -192,6 +192,24 @@ demote = number`{val}
 length : {n, a, b} (fin n, Literal n b) => [n]a -> b
 length _ = `n
 
+/**
+ * Compare the given numeric types, returing
+ * 'True' when 'm == n' and 'False' otherwise.
+ */
+primitive compareEq : { m : #, n : # } Bit
+
+/**
+ * Compare the given numeric types, returning
+ * 'True' when 'm < n' and 'False' otherwise.
+ */
+primitive compareLt : { m : #, n : # } Bit
+
+/**
+ * Compare the given numeric types, returning
+ * 'True' when 'm <= n' and 'False' otherwise.
+ */
+primitive compareLeq : { m : #, n : # } Bit
+
 /**
```

Comment:
In the interest of keeping the number of primitives down, I wonder if it might be better to add a single `demoteNumeric` primitve?   Perhaps something like this:
```
demoteNumeric : { m : # } => (Bool,Integer)
```
The first component indicates if the number is infinite, so finite numbers would be of the form `(false,n)` and infinity would be `(true,0)`.   I guess it would be prettier to expose `InfNat`, but I am not sure it is worth the extra work.

Reply:
I was envisioning a `demoteConstraint` that takes logic about `inf` into account.

Reply:
Also, I was able to get `isPrefix` to work, so perhaps we only need a narrow solution for cases like `coerceSize`?

```
isPrefix : {m, n, a} (fin n, Eq a) => [m]a -> [n]a -> Bit
isPrefix xs x =
    if `(min n m) < `n then False
    else and [ xi == xsi | xi <- x | xsi <- xs ]
```

Reply:
Oh, that trick with `min` is interesting.  We can probably reduce everything we want to do down to just the `compareEq` primitive with clever uses of `min` and `max`.  Or, Iavor's idea would also work.

Reply:
Hm, `demoteConstraint` as a primitive seems tricky, as it is not clear what would be its type.  I think for now we should go with a simpler solution, and the "proper" fix would be to add a GADT style language construct for examining numeric types.